### PR TITLE
librbd/cache: Refactor common code for RWL and SSD

### PR DESCRIPTION
**Pull request by MahatiC**
... from AbstractWriteLog class

Signed-off-by: Lisa Li <xiaoyan.li@intel.com>
Signed-off-by: Mahati Chamarthy <mahati.chamarthy@intel.com>
Signed-off-by: Changcheng Liu <changcheng.liu@intel.com>

## Checklist
- [ ] References tracker ticket
- [ ] Updates documentation if necessary
- [ ] Includes tests for new functionality or reproducer for bug

---

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`

</details>

### CODE REVIEW

**Comment by MahatiC** [comment]
@dillaman This is the next PR that refactors the code for RWL and SSD. Please review.

File: src/librbd/cache/pwl/AbstractWriteLog.h [hunk 1]
```diff
@@ -104,12 +106,11 @@ class AbstractWriteLog {
   CephContext * get_context();
   void release_guarded_request(BlockGuardCell *cell);
   void release_write_lanes(C_BlockIORequestT *req);
-  bool alloc_resources(C_BlockIORequestT *req);
-  template <typename V>
-  void flush_pmem_buffer(V& ops);
+  virtual bool alloc_resources(C_BlockIORequestT *req) = 0;
+  virtual void setup_schedule_append(pwl::GenericLogOperationsVector &ops,
+      bool do_early_flush) = 0;
```
Comment (dillaman) [inline src/librbd/cache/pwl/AbstractWriteLog.h:111]: Nit: align the params indenting

File: src/librbd/cache/pwl/AbstractWriteLog.h [hunk 2]
```diff
@@ -33,6 +34,7 @@ class SyncPointLogEntry;
 class GenericWriteLogEntry;
 class WriteLogEntry;
 class GenericLogEntry;
+struct WriteLogPmemEntry;
```
Comment (dillaman) [inline src/librbd/cache/pwl/AbstractWriteLog.h:37]: Nit: while here, can we rearrange these in alphabetical order?

File: src/librbd/cache/pwl/AbstractWriteLog.h [hunk 3]
```diff
@@ -271,19 +271,23 @@ class AbstractWriteLog {
 
   void pwl_init(Context *on_finish, pwl::DeferredContexts &later);
   void update_image_cache_state(Context *on_finish);
-  void load_existing_entries(pwl::DeferredContexts &later);
   void wake_up();
-  void process_work();
 
+  void update_entries(std::shared_ptr<pwl::GenericLogEntry> log_entry,
```
Comment (dillaman) [inline src/librbd/cache/pwl/AbstractWriteLog.h:276]: Nit: same comment here and below re: aligning the params. Either pull them all down to the line after the method name or align all params on subsequent lines by the indentation of the first parameter.

File: src/librbd/cache/pwl/ReplicatedWriteLog.h [hunk 4]
```diff
@@ -50,6 +50,43 @@ class ReplicatedWriteLog : public AbstractWriteLog<ImageCtxT> {
...
+  using AbstractWriteLog<ImageCtxT>::m_image_ctx;
+  using AbstractWriteLog<ImageCtxT>::m_perfcounter;
+  using AbstractWriteLog<ImageCtxT>::m_ops_to_flush;
+  using AbstractWriteLog<ImageCtxT>::m_cache_state;
+  using AbstractWriteLog<ImageCtxT>::m_first_free_entry;
+  using AbstractWriteLog<ImageCtxT>::m_first_valid_entry;
+
+  PMEMobjpool *m_log_pool = nullptr;
+
+  void process_work() override;
```
Comment (dillaman) [inline src/librbd/cache/pwl/ReplicatedWriteLog.h:64]: Nit: these overrides are technically part of the protected section.

File: src/librbd/cache/pwl/ReplicatedWriteLog.h [hunk 5]
```diff
@@ -50,6 +50,43 @@ class ReplicatedWriteLog : public AbstractWriteLog<ImageCtxT> {
   using C_WriteSameRequestT = pwl::C_WriteSameRequest<This>;
   using C_CompAndWriteRequestT = pwl::C_CompAndWriteRequest<This>;
 
+  using AbstractWriteLog<ImageCtxT>::m_lock;
```
Comment (dillaman) [inline src/librbd/cache/pwl/ReplicatedWriteLog.h:53]: Nit: same comment here -- if these are getting pulled from the protected section of `AbstractWriteLog`, you really can't reduce their visibility in the private section.

**Review by dillaman — COMMENTED** [review]
Nit: can you also provide a better dividing line between the protected and private sections of `AbstractWriteLog`? It seems like there are a lot of things that should be private that are just in the default protected section.

File: src/librbd/cache/pwl/AbstractWriteLog.h (hunk 1 above)
Comment (MahatiC) [reply src/librbd/cache/pwl/AbstractWriteLog.h]: Done

File: src/librbd/cache/pwl/AbstractWriteLog.h (hunk 2 above)
Comment (MahatiC) [reply src/librbd/cache/pwl/AbstractWriteLog.h:37]: Done

File: src/librbd/cache/pwl/AbstractWriteLog.h (hunk 3 above)
Comment (MahatiC) [reply src/librbd/cache/pwl/AbstractWriteLog.h]: Done

File: src/librbd/cache/pwl/ReplicatedWriteLog.h (hunk 4 above)
Comment (MahatiC) [reply src/librbd/cache/pwl/ReplicatedWriteLog.h:73]: Done

File: src/librbd/cache/pwl/ReplicatedWriteLog.h (hunk 5 above)
Comment (MahatiC) [reply src/librbd/cache/pwl/ReplicatedWriteLog.h:64]: Done

File: src/librbd/cache/pwl/AbstractWriteLog.h [hunk 6]
```diff
@@ -205,29 +279,20 @@ class AbstractWriteLog {
    * bufs. Hold a write lock to prevent readers from being added (e.g. when
    * removing log entrys from the map). No lock required to remove readers. */
   mutable RWLock m_entry_reader_lock;
-  /* Hold m_deferred_dispatch_lock while consuming from m_deferred_ios. */
-  mutable ceph::mutex m_deferred_dispatch_lock;
+  //mutable ceph::mutex m_deferred_dispatch_lock; (declared above)
```
Comment (dillaman) [inline src/librbd/cache/pwl/AbstractWriteLog.h:282]: Nit: drop

File: src/librbd/cache/pwl/AbstractWriteLog.cc [hunk 7]
```diff
@@ -396,48 +348,43 @@ void AbstractWriteLog<I>::load_existing_entries(DeferredContexts &later) {
     } else if (pmem_entry->is_write()) {
       ldout(m_image_ctx.cct, 20) << "Entry " << entry_index
                                  << " is a write. pmem_entry=[" << *pmem_entry << "]" << dendl;
-      auto write_entry =
+      auto write_entry =
```
Comment (dillaman) [inline src/librbd/cache/pwl/AbstractWriteLog.cc:351]: Nit: why all the whitespace added at the end of these lines (here and below)?

File: src/librbd/cache/pwl/AbstractWriteLog.cc [hunk 8]
```diff
@@ -1338,62 +1150,39 @@ void AbstractWriteLog<I>::release_guarded_request(BlockGuardCell *released_cell)
...
-        ops_remain = true; /* Always check again before leaving */
-        ldout(m_image_ctx.cct, 20) << "appending " << ops.size() << ", "
-                                   << m_ops_to_append.size() << " remain" << dendl;
-      } else {
-        ops_remain = false;
-        if (appending) {
-          appending = false;
-          m_appending = false;
-        }
+  const unsigned long int OPS_APPENDED = isRWL ? MAX_ALLOC_PER_TRANSACTION : MAX_WRITES_PER_SYNC_POINT;
```
Comment (dillaman) [inline src/librbd/cache/pwl/AbstractWriteLog.cc:1157]: Nit: line wrap after 80 characters

File: src/librbd/cache/pwl/ReplicatedWriteLog.cc [hunk 9]
```diff
@@ -27,14 +30,876 @@
...
+            ldout(cct, 20) << "Freeing " << entry->ram_entry.write_data.oid.pool_uuid_lo
+                           << "." << entry->ram_entry.write_data.oid.off << dendl;
+            TX_FREE(entry->ram_entry.write_data);
+          } else {
+            ldout(cct, 20) << "Retiring non-write: " << *entry << dendl;
+          }
+        }
+      } TX_ONCOMMIT {
+      } TX_ONABORT {
+        lderr(cct) << "failed to commit free of" << retiring_entries.size() << " log entries (" << this->m_log_pool_name << ")" << dendl;
```
Comment (dillaman) [inline src/librbd/cache/pwl/ReplicatedWriteLog.cc:506]: Nit: line wrap

File: src/librbd/cache/pwl/ReplicatedWriteLog.cc [hunk 10]
```diff
@@ -27,14 +30,876 @@
...
+      ldout(m_image_ctx.cct, 10) << "alloc_fail=" << this->m_alloc_failed_since_retire
+                                 << ", allocated > high_water="
+                                 << (this->m_bytes_allocated > high_water_bytes)
+                                 << ", allocated_entries > high_water="
+                                 << (m_log_entries.size() > high_water_entries)
+                                 << dendl;
+      while (this->m_alloc_failed_since_retire || this->m_invalidating ||
+            (this->m_bytes_allocated > high_water_bytes) ||
+            (m_log_entries.size() > high_water_entries) ||
+            (((this->m_bytes_allocated > low_water_bytes) || (m_log_entries.size() > low_water_entries)) &&
```
Comment (dillaman) [inline src/librbd/cache/pwl/ReplicatedWriteLog.cc:749]: Nit: line wrap

File: src/librbd/cache/pwl/AbstractWriteLog.cc (hunk 7 above)
Comment (MahatiC) [reply src/librbd/cache/pwl/AbstractWriteLog.cc]: sorry, that was a mistake. Cleaned up now.

File: src/librbd/cache/pwl/AbstractWriteLog.h (hunk 6 above)
Comment (MahatiC) [reply src/librbd/cache/pwl/AbstractWriteLog.h]: Done

File: src/librbd/cache/pwl/AbstractWriteLog.cc (hunk 8 above)
Comment (MahatiC) [reply src/librbd/cache/pwl/AbstractWriteLog.cc]: Done

File: src/librbd/cache/pwl/ReplicatedWriteLog.cc (hunk 9 above)
Comment (MahatiC) [reply src/librbd/cache/pwl/ReplicatedWriteLog.cc]: Done

File: src/librbd/cache/pwl/ReplicatedWriteLog.cc (hunk 10 above)
Comment (MahatiC) [reply src/librbd/cache/pwl/ReplicatedWriteLog.cc]: Done

**Review by dillaman — APPROVED** [review]
lgtm

**Pull request merged**